### PR TITLE
Resolver redirects

### LINKED ISSUE #47: Cache redirects
Need to support cache redirects.

Motivation is to support queries like:

```graphql
query getAUser($id: ID!) {
  user(id: $id) { ... }
}
```

and return the entity with that id, no matter where it was retrieved from

### PR DESCRIPTION
**Pull request by nevir**
Resolves #47 (also see [design doc](https://docs.google.com/document/d/1x2tQWTJbchfoU8i56NTxwsnqnpDJMQvPVKQ5JiHbSco/))

Note that this only works for parameterized fields, and only allows you to point to entities

### CODE REVIEW

File: test/unit/operations/read/redirects.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      });
+
+      const { result, complete } = read(context, { ...rootRedirectQuery, variables: { id: 1 } }, snapshot);
+      expect((result as any).thing).to.deep.eq({ __typename: 'Thing', id: 111, name: 'Other One' });
+      expect(complete).to.eq(true);
+    });
+
+    it(`supports redirects to nowhere`, () => {
+      const { result, complete } = read(context, { ...rootRedirectQuery, variables: { id: 123 } }, baseSnapshot);
+      expect((result as any).thing).to.deep.eq(undefined);
```
Comment (yuit) [inline test/unit/operations/read/redirects.ts]: Should this be `null` instead?

File: src/context/CacheContext.ts [hunk 2]
```diff
@@ -25,6 +25,17 @@ export namespace CacheContext {
   }
 
   /**
+   * Expected to return an EntityId or undefined, but we loosen the restrictions
```
Comment (yuit) [inline src/context/CacheContext.ts]: We should document here that that this redirect only from parameterized field to Entity only

File: src/context/CacheContext.ts [hunk 3]
```diff
@@ -66,6 +77,12 @@ export namespace CacheContext {
      * Defaults to true unless process.env.NODE_ENV === 'production'
      */
     freeze?: boolean;
+
+    /**
+     * Parameterized fields that should redirect to other locations in the cache
```
Comment (yuit) [inline src/context/CacheContext.ts]: I think we should make sure that "other locations" is strictly an entity

File: test/unit/operations/read/redirects.ts (hunk 1 above)
Comment (nevir) [reply test/unit/operations/read/redirects.ts]: Nope!  `undefined` indicates the value hasn't been populated in the cache - `null` would indicate that we _have_ retrieved it, but it's just not present (on the server)

**Pull request merged**